pkg/kubernetes/reconcile.go

```diff
@@ -14,142 +14,132 @@ import (
+	return fmt.Sprintf("recursion did not resolve in a valid Kubernetes object. "+
+		" In path path `%s` found key `%s` of type `%T` instead.",
+		e.path, e.key, e.primitive)
+}
 
-	// assert for map[string]interface{} (also aliased objx.Map)
-	if m, ok := deep.(objx.Map); ok {
-		deep = map[string]interface{}(m)
-	}
-	deep, ok := deep.(map[string]interface{})
-	if !ok {
-		return fmt.Errorf("deep has unexpected type %T @ %s", deep, path)
+// walkJSON recurses into either a map or list, returning a list of all objects that look
+// like kubernetes resources. We support resources at an arbitrary level of nesting, and
+// return an error if any leaf nodes f
+//
+// Handling the different types is quite gross, so we split this method into a generic
+// walkJSON, and then walkObj/walkList to handle the two different types of collection we
+// support.
+func walkJSON(ptr interface{}, paths ...string) (manifest.List, error) {
+	switch v := ptr.(type) {
+	case map[string]interface{}:
+		return walkObj(v, paths...)
+	case []interface{}:
+		return walkList(v, paths...)
 	}
 
-	// already flat?
-	r := objx.New(deep)
+	key := paths[len(paths)-1]
+	path := "." + strings.Join(paths[:len(paths)-1], ".")
+
+	return nil, ErrorPrimitiveReached{path: path, key: key, primitive: ptr}
+}
+
+//		case d == nil: // result from false if condition in Jsonnet
+func walkObj(obj objx.Map, paths ...string) (manifest.List, error) {
+	obj = obj.Exclude([]string{"__ksonnet"}) // remove our private ksonnet field
 
-	if r.Has("apiVersion") && r.Has("kind") {
-		extracted[path.Full()] = deep.(map[string]interface{})
-		return nil
+	// This looks like a kubernetes manifest, so make one and return it
+	if isKubernetesManifest(obj) {
+		m := manifest.Manifest(obj.Value().MSI())
+
+		return manifest.List{m}, nil
 	}
 
-	// walk it
-	for key, d := range deep.(map[string]interface{}) {
-		switch {
-		case key == "__ksonnet": // exclude ksonnet object
-			continue
-		case d == nil: // result from false if condition in Jsonnet
+	manifests := manifest.List{}
+	for key, value := range obj {
+		if value == nil { // result from false if condition in Jsonnet
 			continue
 		}
-
-		path := append(path, key)
-
-		if _, ok := d.(map[string]interface{}); !ok {
-			return ErrorPrimitiveReached{path.Base(), key, d}
+		children, err := walkJSON(value, append(paths, key)...)
+		if err != nil {
+			return nil, err
 		}
 
-		m := objx.New(d)
-		if m.Has("apiVersion") && m.Has("kind") {
-			mf, err := manifest.NewFromObj(m)
-			if err != nil {
-				return err.(*manifest.SchemaError).WithName(path.Full())
-			}
-			extracted[path.Full()] = mf
-		} else {
-			if err := walkJSON(m, extracted, path); err != nil {
-				return err
-			}
-		}
+		manifests = append(manifests, children...)
 	}
-	return nil
-}
-
-type trace []string
 
-func (t trace) Full() string {
-	return "." + strings.Join(t, ".")
+	return manifests, nil
 }
 
-func (t trace) Base() string {
-	if len(t) > 0 {
-		t = t[:len(t)-1]
```

Comment:
Please stick to this pattern.

I think using this reads much cleaner than the flat-slice approach, because there are proper helper methods available. Also appending to the path is just a matter of `append(...)`. It might be worth adding `push` and `pushIndex` (for arrays) as well